docs/application/native/guides/machine-learning/machine-learning-training.md

```diff
@@ -0,0 +1,475 @@
+                                                    "epsilon=1e-7", NULL);
+status = ml_train_model_set_optimizer(model, optimizer);
+```
+
+### Set a dataset
+
+There are two ways to create a dataset. One is from a file, and another one is from a callback.
+In either fashion, you need to provide streams of tensor data and one-hot-encoded labels.
+
+One-hot-encoded label is a group of values with only one value representing the label,usually one bit high(1), and rests representing others, normally one bit low(0).
+
+#### Set a dataset from a file
+
+To create a `dataset` from a file, a training set and a label set must be provided. A validation set and a test set can be provided if needed.
+
+A label set is a file that has actual labels each in a line. For example:
+
+```
+airplane
+automobile
+bird
+cat
+deer
+dog
+frog
+horse
+ship
+truck
+```
+
+Each dataset except for the label set must contain raw float array data and one-hot-encoded labels:
+
+```
+[float array for data1][label(on-hot encoded) for data1][float array][label]...
+```
+
+After preparing the label set and dataset, create `dataset` as follows:
+
+```c
+int status = ML_ERROR_NONE;
+ml_train_dataset_h dataset;
+
+char *res_path = app_get_resource_path();
+char train_path[1024];
+char label_prop[1024];
+
+snprintf(train_path, sizeof(train_path), "%s/training.dat", res_path);
+snprintf(label_prop, sizeof(label_prop), "label_data=%s/label.dat", res_path);
+free(res_path);
+
+status = ml_train_dataset_create_with_file(&dataset, train_path);
+if(status != ML_ERROR_NONE) {
+  // handle error
+}
+
+status = ml_train_dataset_set_property(dataset, label_prop,
+                                   "buffer_size=9", NULL);
+status = ml_train_model_set_dataset(model, dataset);
+```
+
+`buffer_size` can be set for dataset.
+
+#### Set a dataset from a generator
+
+Creating a dataset from a generator function is also possible.
+
+1. Prepare a callback function:
+
+    ```c
+    /**
+    * @brief      get data which size is batch for train
+    * @param[out] data
+    * @param[out] label
+    * @param[out] last if the data is finished
+    * @param[in] user_data private data for the callback
+    * @retval status for handling error
+    */
+    int get_train_data(float **data, float **label, bool *last, void *user_data) {
+      /* code that fills data, label and last */
+      return ML_ERROR_NONE;
+    }
+    ```
+
+2. Create a dataset from the callback function:
+
+    ```c
+    int status = ML_ERROR_NONE;
+    ml_train_dataset_h dataset;
+
+    // validation and test callback can be omitted.
+    status = ml_train_dataset_create_with_generator(&dataset, get_train_data,
+                                                    NULL, NULL);
+    ```
+
+## Compile the model
+
+Compiling a model finalizes the model with loss.
+Once compiled, any modification to the properties of the model is restricted.
+Adding layers or changing the optimizer/dataset of the model is not permitted neither:
```

Comment:
Suggest rephraing to:

Once compiled, any modification to the properties of the model is restricted.
Adding layers or changing the optimizer or dataset of the model is not permitted either:

Reply:
Done